cmd/clusterctl/cmd/delete.go

```diff
@@ -17,15 +17,82 @@ limitations under the License.
 package cmd
 
 import (
+	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 )
 
+type deleteOptions struct {
+	kubeconfig           string
+	forceDeleteNamespace bool
+	forceDeleteCRD       bool
+	deleteAll            bool
+}
+
+var dd = &deleteOptions{}
+
 var deleteCmd = &cobra.Command{
-	Use:   "delete",
-	Short: "Delete a cluster API resource",
-	Long:  `Delete a cluster API resource with one command`,
+	Use:   "delete [providers]",
+	Short: "Deletes one or more providers from the management cluster",
+	Long: LongDesc(`
+		Deletes one or more providers from the management cluster.`),
+
+	Example: Examples(`
+		# Deletes the AWS provider
+		# Please note that this imply the deletion of everything all the provider components except the 
+		# hosting namespace and the CRDs.
+		clusterctl delete aws
+ 
+		# Deletes the instance of the AWS provider hosted in the "foo" namespace
+		# Please note that, in case there are more than one instances of the AWS provider installed in the cluster,
+		# the provider components shared across providers (e.g. ClusterRoles), are not deleted in order to preserve
+		# the functioning of the remaining instances.
+		clusterctl delete foo/aws
+
+		# Deletes all the providers
+		# Important! As a consequence of this operation all the corresponding resources on target clouds
+		# are "orphaned" and thus there may be ongoing costs incurred as a result of this.
+		clusterctl delete --all
+
+		# Delete the AWS provider and related CRDs. Please note that this forces deletion of 
+		# all the related objects (e.g. AWSClusters, AWSMachines etc.).
+		# Important! As a consequence of this operation all the corresponding resources on AWS cloud
+		# are "orphaned" and thus there may be ongoing costs incurred as a result of this.
+		clusterctl delete aws --delete-crd
+
+		# Delete the AWS provider and its hosting Namespace. Please note that this forces deletion of 
+		# all objects existing in the namespace. 
+		# Important! As a consequence of this operation all the corresponding resources on AWS cloud
+		# are "orphaned" and thus there may be ongoing costs incurred as a result of this.
+		clusterctl delete aws --delete-namespace
```

Comment:
I am trying to understand the pros/cons of providing namespace deletion through clusterctl.
A few questions raised if providing through clusterctl instead of relying on kubectl:
1) Not consistent with init, clusterctl init does not handle namespace creation, or maybe i missed?
2) Namespace Creation/Deletion require elevated privilege, usually handle by different persona in multi-tenancy scenarios

Reply:
> 1. Not consistent with init, clusterctl init does not handle namespace creation, or maybe i missed?

AFAIK all the YAML for installing a provider does include creating a namespace (`capi-system`, `capa-system`, `capv-system` etc.). `clusterctl init` build on top of this.

> 1. Namespace Creation/Deletion require elevated privilege, usually handle by different persona in multi-tenancy scenarios

Provider installation and or deletion requires elevated privileges not only for namespaces, but also for ClusterRoles, Roles, and CRDs. TBH I see it not easy to think this operation to be delegated to someone different than the cluster-admin, but I would like also to hear some other opinion about this.
BTW, I fully agree with you that namespace deletion requires extra care and this is why I'm proposing it requires a separate flag to be explicitly set.

Reply:
The mixing of two different user personas (admin of management cluster vs workload cluster admin creating a cluster using an existing management cluster) is part of the reason why I think there will be continued confusion over the purpose, scope, and use cases for `clusterctl`

Reply:
IMO there is no risk of persona mismatch for the delete command, subject of this thread because this command is in charge of the admin of management cluster.

If instead, we consider other commands like kubeadm config, I got your point. 
From my PoV those are the reason why this should not be a blocker at least in the short term:
- a single tool simplifies the day 1 user experience, and this was one of the main goals of this redesign
- there are already other tools in the ecosystem that targets different personas, most notably. kubectl. I'm confident the user should not be so confused if clusterctl does this as well.
- we are providing a low-level clusterctl library, so, if necessary, a different command for each persona could be easily created at any time.

Reply:
I propose we proceed - this is just a scaffolding. We will have room to iterate.